### DISCUSSION TITLE
`suppressHydrationWarning` not working as expected for app router

### DISCUSSION THREAD

**Discussion by aronwoost** [vercel/next.js#75890] [App Router] [answered]
### Link to the code that reproduces this issue

https://github.com/aronwoost/nextjs-suppress-warning-issue

### To Reproduce

1. `npm install`
2. `npm run dev`
3. Open http://localhost:3000/ in browser

### Current vs. Expected behavior

Actual result: "Hello Server" is displayed  
Expected result: "Hello Client" is displayed

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 20.6.0: Thu Jul  6 22:12:47 PDT 2023; root:xnu-7195.141.49.702.12~1/RELEASE_ARM64_T8101
Binaries:
  Node: 18.17.0
  npm: 9.6.7
  Yarn: 1.22.5
  pnpm: N/A
Relevant Packages:
  next: 14.1.0
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

This is the code in question:

```jsx
"use client";

import styles from "./page.module.css";

export default function Home() {
  let helloLabel = "";

  if (typeof window !== "undefined") {
    helloLabel = "Client";
  } else {
    helloLabel = "Server";
  }

  return (
    <main className={styles.main}>
      <h1 suppressHydrationWarning>Hello {helloLabel}</h1>
    </main>
  );
}
```

`suppressHydrationWarning` does suppress the warning, but it does also prevents re-rendering. Without the `suppressHydrationWarning` prop, the text is correctly displayed but (of course) the warning is printed in the devtools console.

**Comment by kostalexis89** [comment]
I have the same issue. Hope we get a fix soon.

**Comment by aronwoost** [comment]
Issue persists with next@14.1

**Comment by sergio-milu** [comment]
same issue here, it would be nice to now if this is the expected behaviour on next or if it is a bug

**Comment by Vsion** [comment]
+1

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by sebmarkbage** [comment]
It's expected behavior in all React 18+.

**Comment by aronwoost** [comment]
Can you point to documentation/change log? 

I only find this: https://react.dev/reference/react-dom/client/hydrateRoot#suppressing-unavoidable-hydration-mismatch-errors (and have been using `suppressHydrationWarning` for date related issues every since)

**Comment by aronwoost** [comment]
Also - even with React 18 - the Page Router still have the old behavior.

**Comment by capaj** [comment]
@sebmarkbage how come?
Are you saying that putting `suppressHydrationWarning` changes how the JSX element renders?
I would expect it only hides the warning and keeps everything else running the exact same way.

**Comment by felipedeboni** [comment]
Just faced this one, we have an UTC datetime, which server prints on it's local time and on browser it will render on local user's time. 

Using suppressHydrationWarning prevents the datetime to display in local time and also does nothing about the hydration errors.

**Comment by d3vAdv3ntur3s** [comment]
Hit this also as far as I can tell... Using a build time SSG dynamic route, that renders a client side component, e.g.

_blog/[id]/page.tsx_ - Leveraging `generateStaticParams` function for dynamic SSG building of the page
```
return (
    <>
      <h1>{blog.name}</h1>
      <div suppressHydrationWarning={true}> 
        <BlogClient {...blog} />
      </div>
    </>
```
_blog/[id]/_client.tsx_  - Exports client only component BlogClient. (NOTE: tried adding the supress at the root level in the client component as quick test, even though that's client rendered, but no change either) 

Next in dev mode suggests: https://nextjs.org/docs/messages/react-hydration-error#solution-3-using-suppresshydrationwarning 

As discussed above in this thread, it has had no effect, the warnings persist.

**Comment by aronwoost** [comment]
@d3vAdv3ntur3s Be aware that `suppressHydrationWarning` only works one level deep and only with text content ([source](https://react.dev/reference/react-dom/client/hydrateRoot#suppressing-unavoidable-hydration-mismatch-errors)), so not sure if it would work with your example.

**Comment by unhyif** [comment]
I'm experiencing the same issue.
Without `suppressHydrationWarning`, the output is as expected but I have the hydration error.
With `suppressHydrationWarning`, the output isn't as expected but I have no error.

```ts
'use client';

import { updatedDateStyle } from './HouseTableUpdatedDate.css';

interface Props {
  updatedAt: number;
}

const HouseTableUpdatedDate = (props: Props) => {
  const { updatedAt } = props;

  const updatedDate = new Intl.DateTimeFormat('ko', {
    dateStyle: 'full',
    timeStyle: 'medium',
  }).format(updatedAt);

  return <p className={updatedDateStyle}>{updatedDate}</p>;
};

export default HouseTableUpdatedDate;
```

I'm on next 14.2.2

**Comment by shaaandi** [comment]
Any update here 🤔  - same issue and wondering if it will be fixed soon or need to do something else 👀

**Comment by AlparslanAbdikoglu** [comment]
I just wont use app router until the mean time

**Comment by maslowivan** [comment]
Issue still exists in 14.2.13

_Edit by maintainer bot: Comment was **automatically** minimized because it was considered unhelpful. (If you think this was by mistake, let us know). Please only comment if it adds context to the issue. If you want to express that you have the same problem, use the upvote 👍 on the issue description or subscribe to the issue for updates. Thanks!_

**Comment by brianonchain** [comment]
> @sebmarkbage how come? Are you saying that putting `suppressHydrationWarning` changes how the JSX element renders? I would expect it only hides the warning and keeps everything else running the exact same way.

I believe it does. Using "suppressHydrationWarning" is giving me different results in the rendered DOM, it doesn't simply just suppress warnings in Chrome dev tools.

**Comment by WilliamBlais** [comment]
Is there any workaround except not using suppressHydrationWarning?

**Comment by peyronoscar** [comment]
<img width="965" alt="Screenshot 2024-12-05 at 03 10 21" src="https://github.com/user-attachments/assets/631601ff-fa21-4f57-87db-4ddac9102622">


"next": "15.0.3",
"react": "19.0.0-rc-66855b96-20241106",
"react-dom": "19.0.0-rc-66855b96-20241106"

**Comment by evheniydan** [comment]
Updates? NextJS 15, same issue. HTML with suppressHydrationWarning={true} is not disabling hydration warnings.

**Comment by amiftachulh** [comment]
I use Next.js v15.1.4 and Chakra UI v3.0.0 with Turbopack enabled and still show hydration error although I already give html tag `suppressHyrationWarning`.
![image](https://github.com/user-attachments/assets/3ff3dfb0-0e7a-4978-a172-a6a29395dcf8)
However when I disable Turbopack the hydration error won't appear
![image](https://github.com/user-attachments/assets/f0d7c06a-33c3-42f9-9db2-9bf35dd88b93)

**Comment by oxedom** [comment]
This issue has been quite a headache.

 The only solution that worked for dev was @amiftachulh's suggestion (removing `--turbopack` from `yarn dev`), but unfortunately, deploying to production still throws a client-side error: 

`Application error: a client-side exception has occurred (see the browser console for more information).`

["Turbopack currently only supports next dev and does not support next build"](https://nextjs.org/docs/app/api-reference/turbopack#unsupported-features) so I'm still uncertain where this error is coming from...

**Comment by samcx** [comment, marked as answer]
@aronwoost Can confirm this is the expected behavior, not just in App Router. → https://github.com/facebook/react/pull/26391

Our Documentation should be updated 👀

**Reply by samcx** [reply]
follow up: https://github.com/vercel/next.js/pull/76521

**Comment by foysal-m** [comment]
I am getting this hydration error, and not having a clue yet.

<img width="651" alt="image" src="https://github.com/user-attachments/assets/25baf161-1f40-4cb7-a44e-9ff89abf1720" />

**Reply by BuhayovA** [reply]
try to open an incognito tab and check for the error

[source](https://www.reddit.com/r/nextjs/comments/1gabiqn/comment/ltcfnk7/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button)

**Reply by Misyuk-T** [reply]
have the same problem

**Comment by timothyej** [comment]
Here's a workaround that worked for me:

```jsx
'use client';
import { useEffect, useState } from 'react';

const SuppressHydrationWarning = ({ children }) => {
  const [isClient, setIsClient] = useState(false);

  useEffect(() => {
    setIsClient(true);
  }, []);

  if (isClient) {
    return children;
  }

  return (
    <span suppressHydrationWarning>
      {children}
    </span>
  );
};
```

Example of how to use it:

```jsx
<time dateTime={date.toISOString()}>
  <SuppressHydrationWarning>
    {date.toLocaleTimeString(undefined, { timeStyle: 'short' })}
  </SuppressHydrationWarning>
</time>
```

It works exactly as you'd expect it to, it renders both on the server and the client and suppresses the hydration warning.

**Comment by nikolayemrikh** [comment]
@timothyej it does work, thanks

Btw that's super weird that just "supressing a warning" somehow affects rendering. This issue should be raised

**Comment by nikolayemrikh** [comment]
```tsx
export function SuppressHydrationWarning<C extends ElementType>({
  component: Component,
  ...restProps
}: ComponentPropsWithRef<C> & { component: C }): ReactNode {
  const [isMounted, setIsMounted] = useState(false);

  useEffect(() => {
    setIsMounted(true);
  }, []);

  return <Component suppressHydrationWarning={!isMounted} key={isMounted} {...restProps} />;
}
```

Modified the solution above to make it more generic and type safe

Usage:

```tsx
<SuppressHydrationWarning component="button" onClick={() => {/* */}}>
  content
</SuppressHydrationWarning>
```
